### ISSUE TITLE
[MultiGPU,In-batch negatives, MultipleNegativeRankingLoss] Ability to share negatives across devices

### ISSUE THREAD

**Issue by olivierr42** [UKPLab/sentence-transformers#2831]
Love that SentenceTransformers was now properly integrated to the Trainer API, leveraging a lot of its key features! 

I think something that could help leverage the multi GPU features in the context of the MultipleNegativeRankingLoss would be to have the ability to share the negatives across devices. [FlagEmbedding](https://github.com/FlagOpen/FlagEmbedding/blob/master/FlagEmbedding/baai_general_embedding/finetune/modeling.py) did it in their implementation.

It essentially requires to add this piece of code (this is from the link share above):

```
import torch.distributed as dist
class MultipleNegativesRankingLoss(nn.Module):
    def __init__(self, model: SentenceTransformer, scale: float = 20.0, similarity_fct=util.cos_sim, negatives_cross_device:bool = True) -> None:
        super(MultipleNegativesRankingLoss, self).__init__()
        self.model = model
        self.scale = scale
        self.similarity_fct = similarity_fct
        self.cross_entropy_loss = nn.CrossEntropyLoss()

        self.negatives_cross_device = negatives_cross_device
        if self.negatives_cross_device:
            if not dist.is_initialized():
                raise ValueError('Distributed training has not been initialized for representation all gather.')
            #     logger.info("Run in a single GPU, set negatives_cross_device=False")
            #     self.negatives_cross_device = False
            # else:
            self.process_rank = dist.get_rank()
            self.world_size = dist.get_world_size()
        
    def _dist_gather_tensor(self, t: Optional[torch.Tensor]):
        if t is None:
            return None
        t = t.contiguous()

        all_tensors = [torch.empty_like(t) for _ in range(self.world_size)]
        dist.all_gather(all_tensors, t)

        all_tensors[self.process_rank] = t
        all_tensors = torch.cat(all_tensors, dim=0)

        return all_tensors
        
    def forward(self, sentence_features: Iterable[Dict[str, Tensor]], labels: Tensor) -> Tensor:
        reps = [self.model(sentence_feature)["sentence_embedding"] for sentence_feature in sentence_features]
        embeddings_a = reps[0]
        embeddings_b = torch.cat(reps[1:])

        embeddings_a = self._dist_gather_tensor(embeddings_a)
        embeddings_b = self._dist_gather_tensor(embeddings_b)

        scores = self.similarity_fct(embeddings_a, embeddings_b) * self.scale
        # Example a[i] should match with b[i]
        range_labels = torch.arange(0, scores.size(0), device=scores.device)
        return self.cross_entropy_loss(scores, range_labels)
```
Disclaimer: This snippet is untested.
I am also wondering how that would fit in CachedMultipleNegativeRankingLoss -- I think [GradCache](https://github.com/luyug/GradCache/blob/main/src/grad_cache/loss.py) also implemented it.

**Comment by tomaarsen** [comment]
Hello!

Apologies for the delay, I've been recovering from a surgery this last month.
I must say that I'm a bit hesitant to add cross-device in-batch negatives, because I believe equivalent results can be reached with `CachedMultipleNegativeRankingLoss`. If you want e.g. 2048 in-batch negatives and you have 8 GPUs, then you can use such a distributed MNRL class with 256 batch size per device, but you can also use CachedMNRL with a batch size per device of 2048. I believe these are equivalent, but the former requires cross-device all-gathers of large matrices. My suspicion is that the CachedMNRL approach would be faster, despite CachedMNRL being ~20% slower than (non-distributed) MNRL iirc.

What do you think?

- Tom Aarsen

**Comment by olivierr42** [comment]
That makes a lot of sense actually! It also avoids the handling of possible duplicates across device (since there is no clear way to prevent that inter-batch).

I will get back to you in terms a performance w.r.t. cross-device negatives.

The main reason behind this issue is that the most powerful solutions of late seemed to rely on enormous batch sizes (>10k), but they never mention the use of caching in their papers, so I assumed they were sharing them across device for training speed reasons. 

Thank you.

**Comment by tomaarsen** [comment]
> I tried running it with the CachedGIST loss, but there seems to be an issue with the tokenization & DDP with this loss at the moment (I think it will be fixed in a future release, as far as I saw).

Indeed, this *should* be resolved in the next release via #2775.

Thanks for experimenting! I imagine that both approaches work, and I'm glad to hear that the CachedMNRL version was fast. I think we can close this issue now :)

- Tom Aarsen
